solidity-bindgen-macros/src/abi_gen.rs

```diff
@@ -0,0 +1,159 @@
+    let abi_str = String::from_utf8(bytes).unwrap();
+
+    let struct_name = Ident::new(name.as_str(), span);
+
+    let fns = abis.iter().map(|abi| fn_from_abi(abi, span));
+
+    quote! {
+        pub struct #struct_name {
+            contract: ::solidity_bindgen::internal::ContractWrapper,
+        }
+
+        impl #struct_name {
+            pub fn new(address: ::web3::types::Address, url: &str, event_loop_handle: ::std::sync::Arc<::web3::transports::EventLoopHandle>) -> ::solidity_bindgen::internal::Result<Self> {
+                // Embed ABI into the program
+                let abi = #abi_str;
+
+                // Set up a wrapper so we can make calls
+                let contract = ::solidity_bindgen::internal::ContractWrapper::new(address, abi.as_bytes(), url, event_loop_handle)?;
+                Ok(Self {
+                    contract
+                })
+            }
+
+            #(#fns)*
+        }
+    }
+}
+
+fn param_token_type(param_type: ParamType) -> TokenStream {
+    match param_type {
+        ParamType::Address => quote! { ::web3::types::Address },
+        ParamType::Bytes => quote! { ::std::vec::Vec<u8> },
+        ParamType::Int(size) => match size {
+            256 => quote! { ::solidity_bindgen::internal::Unimplemented },
+            _ => {
+                let name = Ident::new(&format!("i{}", size), Span::call_site());
+                quote! { #name }
+            }
+        },
+        ParamType::Uint(size) => match size {
+            256 => quote! { ::web3::types::U256 },
+            _ => {
+                let name = Ident::new(&format!("u{}", size), Span::call_site());
+                quote! { #name }
+            }
+        },
+        ParamType::Bool => quote! { bool },
+        ParamType::String => quote! { ::std::string::String },
+        ParamType::Array(inner) => {
+            let inner = param_token_type(*inner);
+            quote! { ::std::vec::Vec<#inner> }
+        }
+        ParamType::FixedBytes(len) => quote! { [ u8; #len ] },
+        ParamType::FixedArray(inner, len) => {
+            let inner = param_token_type(*inner);
+            quote! { [#inner; #len] }
+        }
+        ParamType::Tuple(members) => {
+            match members.len() {
+                0 => {
+                    // TODO: Verify that this detokenizes the same
+                    quote! { ::solidity_bindgen::internal::Empty }
+                }
+                _ => {
+                    let members = members.into_iter().map(|member| param_token_type(*member));
+                    quote! { (#(#members,)*) }
+                }
+            }
+        }
+    }
+}
+
+/// Convert some Ethereum ABI type to a Rust type (usually from the web3 namespace)
+fn param_type(type_name: &str) -> TokenStream {
+    param_token_type(Reader::read(type_name).unwrap())
+}
+
+pub fn to_rust_name(eth_name: &str, i: usize) -> String {
+    if eth_name == "" {
+        format!("no_name_provided_{}", i)
```

Comment:
What we've done in graph-cli code generation is distinguish function inputs and outputs, and enumerate them as `input1`, `input2`, ... and `output1`, `output2`, ...

In structs, for unnamed fields, we've done `value1`, `value2` etc.

That could be nice here too. But it's on the nit-picky side, since the generated code will not usually be observed by humans.

Reply:
Even though the generated code may not be observed by humans, it's documentation is. So when you use the IDE you can already see all the parameter names for calling functions, etc.

Reply:
True. Then I would prefer `input1`, `input2`, `output1`, `output2` etc. over `no_name_provided_1` etc. even more.

Reply:
I changed this already. I think you're reviewing an old version.